engine/src/main/java/org/terasology/rendering/dag/ShadowMapNode.java

```diff
@@ -0,0 +1,146 @@
+
+    private Camera playerCamera;
+    // FIXME: remove reference to WorldRendererImpl
+    private WorldRenderer worldRenderer;
+    private BackdropProvider backdropProvider;
+
+
+    // FIXME: unnecessary arguments must be eliminated
+    public ShadowMapNode(Context context, Material shadowMapShader, Camera playerCamera) {
+        this.context = context;
+        this.shadowMapShader = shadowMapShader;
+        this.playerCamera = playerCamera;
+        this.worldRenderer = context.get(WorldRenderer.class);
+        this.backdropProvider = context.get(BackdropProvider.class);
+        this.renderingConfig = context.get(Config.class).getRendering();
+        this.shadowMap = context.get(FrameBuffersManager.class).getFBO("sceneShadowMap");
+        this.renderQueues = context.get(RenderQueuesHelper.class);
+    }
+
+    @Override
+    public void update(float deltaInSeconds) {
+        this.shadowMap = context.get(FrameBuffersManager.class).getFBO("sceneShadowMap");
+
+        // positionShadowMapCamera()
+
+        // Shadows are rendered around the player so...
+        Vector3f lightPosition = new Vector3f(playerCamera.getPosition().x, 0.0f, playerCamera.getPosition().z);
+
+        // Project the shadowMapCamera position to light space and make sure it is only moved in texel steps (avoids flickering when moving the shadowMapCamera)
+        float texelSize = 1.0f / renderingConfig.getShadowMapResolution();
+        texelSize *= 2.0f;
+
+        camera.getViewProjectionMatrix().transformPoint(lightPosition);
+        lightPosition.set(TeraMath.fastFloor(lightPosition.x / texelSize) * texelSize, 0.0f, TeraMath.fastFloor(lightPosition.z / texelSize) * texelSize);
+        camera.getInverseViewProjectionMatrix().transformPoint(lightPosition);
+
+        // ... we position our new shadowMapCamera at the position of the player and move it
+        // quite a bit into the direction of the sun (our main light).
+
+        // Make sure the sun does not move too often since it causes massive shadow flickering (from hell to the max)!
+        float stepSize = 50f;
+        Vector3f sunDirection = backdropProvider.getQuantizedSunDirection(stepSize);
+
+        Vector3f sunPosition = new Vector3f(sunDirection);
+        sunPosition.scale(256.0f + 64.0f);
+        lightPosition.add(sunPosition);
+
+        camera.getPosition().set(lightPosition);
+
+        // and adjust it to look from the sun direction into the direction of our player
+        Vector3f negSunDirection = new Vector3f(sunDirection);
+        negSunDirection.scale(-1.0f);
+
+        camera.getViewingDirection().set(negSunDirection);
+
+        camera.update(deltaInSeconds);
+
+    }
+
+
+    @Override
+    public void process() {
+        PerformanceMonitor.startActivity("Render World (Shadow Map)");
+
+        // preRenderSetupSceneShadowMap
+        shadowMap.bind();
+        setViewportToSizeOf(shadowMap); // TODO: how about shadowMap.setAsViewport() ?
```

Comment:
Hmmmm. Reading shadowMap.setAsViewport() I'd think that something is happening to the object shadowMap rather than the to viewport. Normally a method affects the instance is attached to or the instance is the "agent" that causes something. In this case neither make sense. We could theorethically go for something like shadowMap.setViewportToItsSize() but it's a bit convoluted. I prefer the immediacy of setViewportToSizeOf(anFBO), as it's almost plain English.